File: release/policy_check/signer.py
"""Signer authority and governance transaction checks."""

from typing import Any
import tomllib

from .common import unique_sorted_fingerprints

TYPES = {"bootstrap", "addition", "replacement", "removal-only", "classification-only"}
ROLES = {"maintainer-administrator", "architect-security"}


def authority_document_valid(document: str) -> bool:
    try:
        parsed = tomllib.loads(document)
    except (tomllib.TOMLDecodeError, TypeError):
        return False
    return (
        set(parsed) == {"schema_version", "authorized_primary_fingerprints"}
        and type(parsed["schema_version"]) is int
        and parsed["schema_version"] == 1
        and unique_sorted_fingerprints(parsed["authorized_primary_fingerprints"])
    )


def transaction_valid(case: dict[str, Any]) -> bool:
    required = {"id", "old", "new", "declared_type", "proof", "approvals", "expected"}
    if set(case) != required:
        return False
    old, new = case["old"], case["new"]
    if not unique_sorted_fingerprints(old) or not unique_sorted_fingerprints(new):
        return False
    introduced = sorted(set(new) - set(old))
    removed = sorted(set(old) - set(new))
    actual_type = _transaction_type(old, new, introduced, removed)
    if actual_type is None or case["declared_type"] not in TYPES or case["declared_type"] != actual_type:
        return False
    if not _approvals_valid(case["approvals"]):
        return False
    return _proof_valid(case["proof"], introduced)


def _transaction_type(old: list[str], new: list[str], introduced: list[str], removed: list[str]) -> str | None:
    if not old and introduced and not removed:
        return "bootstrap"
    if introduced and not removed:
        return "addition"
    if introduced and removed:
        return "replacement"
    if removed and not introduced:
        return "removal-only"
    if old == new:
        return "classification-only"
    return None


def _approvals_valid(approvals: Any) -> bool:
    if not isinstance(approvals, list) or len(approvals) != 2:
        return False
    if any(not isinstance(item, dict) or set(item) != {"person", "role"} for item in approvals):
        return False
    people = [item["person"] for item in approvals]
    roles = [item["role"] for item in approvals]
    return (
        all(isinstance(person, str) and person and not person.startswith("automation-") for person in people)
        and len(set(people)) == 2
        and set(roles) == ROLES
    )


def _proof_valid(proof: Any, introduced: list[str]) -> bool:
    if not isinstance(proof, dict) or set(proof) != {"state", "reason", "introduced_signers"}:
        return False
    records = proof["introduced_signers"]
    if not isinstance(records, list):
        return False
    if introduced:
        if proof["state"] != "verified" or proof["reason"] is not None:
            return False
        fingerprints = [record.get("primary_fingerprint") for record in records if isinstance(record, dict)]
        return (
            all(
                set(record) == {"primary_fingerprint", "verifier_result"}
                and isinstance(record["verifier_result"], str)
                and bool(record["verifier_result"])
                for record in records
                if isinstance(record, dict)
            )
            and len(records) == len(fingerprints)
            and fingerprints == introduced
        )
    return (
        proof["state"] == "not-applicable"
        and isinstance(proof["reason"], str)
        and bool(proof["reason"])
        and records == []
    )

File: tools/release-policy/src/policy/signer.rs
use std::collections::BTreeSet;

use serde::Deserialize;
use serde_json::Value;

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct AuthorityDocument {
    schema_version: u64,
    authorized_primary_fingerprints: Vec<String>,
}

pub(super) fn authority_valid(bytes: &[u8]) -> bool {
    let Ok(document) = std::str::from_utf8(bytes) else {
        return false;
    };
    let Ok(value) = toml::from_str::<AuthorityDocument>(document) else {
        return false;
    };
    if value.schema_version != 1 {
        return false;
    }
    let fingerprints: Vec<&str> = value
        .authorized_primary_fingerprints
        .iter()
        .map(String::as_str)
        .collect();
    fingerprints_valid(&fingerprints)
}

pub(super) fn transaction_reason(case: &Value) -> Option<&'static str> {
    let Some(object) = case.as_object() else {
        return Some("governance-transition-or-proof");
    };
    let required = [
        "id",
        "old",
        "new",
        "declared_type",
        "proof",
        "approvals",
        "expected",
    ];
    if object.len() != required.len() || !required.iter().all(|name| object.contains_key(*name)) {
        return Some("governance-transition-or-proof");
    }
    let Some(old) = string_array(object.get("old")) else {
        return Some("governance-transition-or-proof");
    };
    let Some(new) = string_array(object.get("new")) else {
        return Some("governance-transition-or-proof");
    };
    if !fingerprints_valid(&old) || !fingerprints_valid(&new) {
        return Some("governance-transition-or-proof");
    }
    let introduced: Vec<&str> = new
        .iter()
        .copied()
        .filter(|item| !old.contains(item))
        .collect();
    let removed: Vec<&str> = old
        .iter()
        .copied()
        .filter(|item| !new.contains(item))
        .collect();
    let actual = transaction_type(&old, &new, &introduced, &removed);
    if actual != object.get("declared_type").and_then(Value::as_str) {
        return Some("governance-transition-or-proof");
    }
    if !approvals_valid(object.get("approvals")) {
        return Some("governance-review-or-hold");
    }
    if !proof_valid(object.get("proof"), &introduced) {
        return Some("governance-transition-or-proof");
    }
    None
}

pub(super) fn fingerprints_valid(values: &[&str]) -> bool {
    let sorted: BTreeSet<&str> = values.iter().copied().collect();
    values.iter().all(|value| fingerprint(value))
        && sorted.len() == values.len()
        && sorted.into_iter().eq(values.iter().copied())
}

pub(super) fn fingerprint(value: &str) -> bool {
    matches!(value.len(), 40 | 64)
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'A'..=b'F').contains(&byte))
}

fn string_array(value: Option<&Value>) -> Option<Vec<&str>> {
    value?
        .as_array()?
        .iter()
        .map(Value::as_str)
        .collect::<Option<Vec<_>>>()
}

fn transaction_type<'a>(
    old: &[&str],
    new: &[&str],
    introduced: &[&str],
    removed: &[&str],
) -> Option<&'a str> {
    if old.is_empty() && !introduced.is_empty() && removed.is_empty() {
        Some("bootstrap")
    } else if !introduced.is_empty() && removed.is_empty() {
        Some("addition")
    } else if !introduced.is_empty() && !removed.is_empty() {
        Some("replacement")
    } else if introduced.is_empty() && !removed.is_empty() {
        Some("removal-only")
    } else if old == new {
        Some("classification-only")
    } else {
        None
    }
}

fn approvals_valid(value: Option<&Value>) -> bool {
    let Some(approvals) = value.and_then(Value::as_array) else {
        return false;
    };
    if approvals.len() != 2 {
        return false;
    }
    let mut people = BTreeSet::new();
    let mut roles = BTreeSet::new();
    for approval in approvals {
        let Some(object) = approval.as_object() else {
            return false;
        };
        if object.len() != 2 {
            return false;
        }
        let Some(person) = object.get("person").and_then(Value::as_str) else {
            return false;
        };
        let Some(role) = object.get("role").and_then(Value::as_str) else {
            return false;
        };
        if person.is_empty() || person.starts_with("automation-") {
            return false;
        }
        people.insert(person);
        roles.insert(role);
    }
    people.len() == 2 && roles == BTreeSet::from(["architect-security", "maintainer-administrator"])
}

fn proof_valid(value: Option<&Value>, introduced: &[&str]) -> bool {
    let Some(proof) = value.and_then(Value::as_object) else {
        return false;
    };
    if proof.len() != 3 {
        return false;
    }
    let Some(records) = proof.get("introduced_signers").and_then(Value::as_array) else {
        return false;
    };
    if introduced.is_empty() {
        return proof.get("state").and_then(Value::as_str) == Some("not-applicable")
            && proof
                .get("reason")
                .and_then(Value::as_str)
                .is_some_and(|reason| !reason.is_empty())
            && records.is_empty();
    }
    if proof.get("state").and_then(Value::as_str) != Some("verified")
        || !proof.get("reason").is_some_and(Value::is_null)
    {
        return false;
    }
    let fingerprints: Option<Vec<&str>> = records
        .iter()
        .map(|record| {
            let object = record.as_object()?;
            if object.len() != 2
                || object
                    .get("verifier_result")
                    .and_then(Value::as_str)
                    .is_none_or(str::is_empty)
            {
                return None;
            }
            object.get("primary_fingerprint").and_then(Value::as_str)
        })
        .collect();
    fingerprints.is_some_and(|values| values == introduced)
}

#[cfg(test)]
#[path = "signer/tests.rs"]
mod tests;
